### PR TITLE
Kernel: Dynamically allocated futexes

### LINKED ISSUE #25497: arbitrary futex pointers
**Is your enhancement proposal related to a problem? Please describe.**
Right now, all `k_futex` objects, even though they can live in user memory, must be known to the kernel at build time.

This is because all `k_futex` pointers found are placed in the system's kernel object table generated by gperf. When a thread later makes a call like `k_futex_wait()`, the address of the futex is looked up in the table, and if found, the returned `struct z_object` has a link to that futex's associated `z_futex_data`, containing that futex's wait queue.

The current implementation was easy to do; we already had all the infrastructure for kernel objects and it was simple enough to add `k_futex` to the set of objects and generate its wait queue. But we can do better!

**Describe the solution you'd like**
We'd like user mode to call `k_futex_wait()` on *any* user-accessible memory location. This is how it works in Linux. We'd like to be able to do this without requiring heap allocations.

The basic idea is to re-implement `k_futex_find_data()`. Stop using the kobject table, and instead maintain a special, separate runtime table of active futexes, i.e. futexes that at least one thread is waiting on. This table is keyed by the futex address. When we go virtual memory later, the table will need to be keyed by address plus whatever VM space the address is from.

On a `k_futex_wait()` syscall:
- Validate the caller has access to the memory location like we currently do
- Look up the futex address in the table. If present, wait on its wait_q_t.
- If not present, create a table entry for it, and then wait, or return -ENOMEM if we ran out of table entries.

On a `k_futex_wake()` syscall:
- Validate the caller has access to the memory location like we currently do
- Look up the futex address in the table. If not found, nobody was waiting. Do nothing.
- Otherwise, wake one or wake all waiters. If this wakes everyone up, remove from the table.

The nice thing is that the maximum number of table entries needed is a function of the number of active threads in the system; worst-case is a scenario of all threads waiting on different futexes.

We don't have a general-purpose runtime hash table data structure in Zephyr for the table. But we can make a red-black tree, and that should work for lookups. We similarly use one already for dynamic kernel objects.

So I'm thinking we can extend `z_futex_data` to look something like:

```
struct z_futex_data {
	/* Queue of pending threads */
	_wait_q_t wait_q;
	/* Protect wait queue */
	struct k_spinlock lock;
	/* Red-black tree node data */
	struct rbnode node;
	/* Address that was waited on */
	void *addr;
}
```

And instantiate a `k_mem_slab` pool of these. The number of entries can either be the number of `k_thread`s detected by the kobject mechanism, or a Kconfig. Probably simplest to start with a Kconfig. There would then be a `struct rbtree` to maintain the table which starts out empty.

This approach should also be applicable to `sys_mutex` which is essentially a special kind of futex backed with a `k_mutex` instead of a `_wait_q_t`.

**Describe alternatives you've considered**
None so far.

### PR DESCRIPTION
**Pull request by RNRNRNR**
Fixes #25497

Rework of futex API implementation to make futex data dynamically allocated.  Replaces kernel object table usage to runtime table of active futexes.
Changes implementation to: 

- k_futex_wait() - Validate the caller has access to the memory location like we currently do. Look up the futex address in the table. If present, wait on its wait_q_t. If not present, create table entry for it, and then wait, or return -ENOMEM if we ran out table entries.

- k_futex_wake()  - Validate the caller has access to the memory location like we currently do. Look up the futex address in the table. If not found, nobody was waiting. Do nothing. Otherwise, wake one or wake all waiters. If this wakes everyone up, remove it from the table.

Also, kernel object type K_OBJ_FUTEX and tests related to it are removed

### CODE REVIEW

**Comment by andyross** [comment]
Just noticing this.  Can we get some background here? What's the benefit being provided?  This seems to me to be adding runtime overhead and dependencies on a heap for not much benefit.  IMHO our futex implementation is actually too heavyweight, and this seems to be moving in the wrong direction.  By definition a thread can be blocked on only one object, and that's the only tracking needed in the kernel.  Can't we roll the lookup of "who is blocked on this address" into the thread objects themselves (e.g. union it into the existing node pointer used in the scheduler).

Short version: futexes should be closer to the scheduler, this seems to be moving them farther away.

**Comment by RNRNRNR** [comment]
@andyross Current implementation of futexes restrict their usage in dynamic primitives that contains sem/mutex. This forces to use kernel implementations of sem/mutex in userspace, which means unconditional syscalls and performance impact.
Here's full background #36228.

> By definition a thread can be blocked on only one object, and that's the only tracking needed in the kernel. Can't we roll the lookup of "who is blocked on this address" into the thread objects themselves (e.g. union it into the existing node pointer used in the schedule)
We can add field "address that thread is waiting on" and get futex data through existing wait_q pointer. But isn't lookup through all threads are more overhead?

File: include/kernel.h [hunk 1]
```diff
@@ -1976,7 +1978,7 @@ struct z_futex_data {
  *		  one of the special values K_NO_WAIT or K_FOREVER.
  * @retval -EACCES Caller does not have read access to futex address.
  * @retval -EAGAIN If the futex value did not match the expected parameter.
- * @retval -EINVAL Futex parameter address not recognized by the kernel.
+ * @retval -ENOMEM Not enought memory to place futex at kernel
```
Comment (peter-mitsis) [inline include/kernel.h:1965]: Typo: enought -> enough

**Comment by chbusold** [comment]
> Just noticing this. Can we get some background here? What's the benefit being provided? This seems to me to be adding runtime overhead and dependencies on a heap for not much benefit. IMHO our futex implementation is actually too heavyweight, and this seems to be moving in the wrong direction. By definition a thread can be blocked on only one object, and that's the only tracking needed in the kernel. Can't we roll the lookup of "who is blocked on this address" into the thread objects themselves (e.g. union it into the existing node pointer used in the scheduler).
> 
> Short version: futexes should be closer to the scheduler, this seems to be moving them farther away.

Hi @andyross, I would like to make progress on #36228. Can you elaborate on your futex proposal? Iiuc, you would like to store the address a thread is waiting on inside the thread control block during `k_futex_wait`. During `k_futex_wake` we then go through all threads, see if any is waiting on this address, and wake them (or the one with highest priority). Is this correct? That means we would not be using `_wait_q_t` here at all, and thus no RB tree for handling a large number of threads efficiently.

**Comment by chbusold** [comment]
We could also implement both and make it a choice, like `CONFIG_WAITQ_SCALABLE`. Potentially with the default depending on the number of threads.
